datajoint/dependencies.py

```diff
@@ -101,4 +101,5 @@ def descendants(self, full_table_name):
         :return: all dependent tables sorted in topological order.  Self is included.
         """
         nodes = nx.algorithms.dag.descendants(self, full_table_name)
-        return [full_table_name] + nx.algorithms.dag.topological_sort(self, nodes)
+        sort = nx.algorithms.dag.topological_sort(self)
+        return [full_table_name] + [s for s in sort if s in nodes]
```

Comment:
alternate/better performing implementation suggestion @eywalker:
create a subgraph from full_table_name and descendants and then sort

Reply:
Based on a test of calling ```descendants``` 10000 times for every class within the graph of 41 classes discussed elsewhere, I was able to get the following results locally:

1) initial code above: ~250 iterations of the subclasses-for-each-classes per second
2) Commented code in patch below: ~125 iterations/sec
3) Live code in patch below: ~450 iterations/sec, however this requires making ```connection``` optional in the ```Dependencies``` constructor

can provide the test code elsewhere if that might be useful.

Patch:
--8<--
```
diff --git a/datajoint/dependencies.py b/datajoint/dependencies.py
index 6aa169b..2c0167f 100644
--- a/datajoint/dependencies.py
+++ b/datajoint/dependencies.py
@@ -8,7 +8,7 @@ class Dependencies(nx.DiGraph):
     """
     The graph of dependencies (foreign keys) between loaded tables.
     """
-    def __init__(self, connection):
+    def __init__(self, connection=None):
         self._conn = connection
         self._node_alias_count = itertools.count()
         super().__init__(self)
@@ -95,7 +95,7 @@ class Dependencies(nx.DiGraph):
         return dict(p[1:3] for p in self.out_edges(table_name, data=True)
                     if primary is None or p[2]['primary'] == primary)
 
-    def descendants(self, full_table_name):
+    def descendants_old(self, full_table_name):
         """
         :param full_table_name:  In form `schema`.`table_name`
         :return: all dependent tables sorted in topological order.  Self is included.
@@ -103,3 +103,18 @@ class Dependencies(nx.DiGraph):
         nodes = nx.algorithms.dag.descendants(self, full_table_name)
         sort = nx.algorithms.dag.topological_sort(self)
         return [full_table_name] + [s for s in sort if s in nodes]
+
+    def descendants(self, full_table_name):
+        """
+        :param full_table_name:  In form `schema`.`table_name`
+        :return: all dependent tables sorted in topological order.  Self is included.
+        """
+
+        # nodes = nx.DiGraph(self).subgraph(
+        #     nx.algorithms.dag.descendants(self, full_table_name))
+
+        nodes = self.subgraph(
+            nx.algorithms.dag.descendants(self, full_table_name))
+
+        return [full_table_name] + list(
+            nx.algorithms.dag.topological_sort(nodes))
```

this implies that a similar modification to ```erd.py``` line 209 discussed in other comments could also improve performance; again, however, this requires making ```connection``` optional in the ```Dependencies``` code. The previous implementation running at ~250 iterations shouldn't be impacted by a similar change since it is not constructing copied instances within the main method body. As a side note, unit tests still pass with ```connnection`` an optional argument.